src/workerd/io/worker-interface.capnp

```diff
@@ -147,6 +186,277 @@ struct Trace @0x8e8d911203762d34 {
+    value :union {
+      bool @2 :Bool;
+      int64 @3 :Int64;
+      uint64 @4 :UInt64;
+      float64 @5 :Float64;
+      text @6 :Text;
+      data @7 :Data;
+    }
+  }
+
+  struct StreamEvent {
+    # A unique identifier used to correlate traces across multiple events
+    # in a single tail session. Typically this will correlate to a top-level
+    # pipeline or specific pipeline stage.
+    id @0 :Text;
+
+    span :group {
+      id @1 :UInt32;
+      parent @2 :UInt32;
+      # Some spans may be transactional, meaning that they represent a single
+      # atomic operation that may succeed or fail as a whole. When the span
+      # event occurs, the span outcome will indicate if the span was successful
+      # or failed. If a transaction span outcome indicates failure, then all events
+      # within that span should be considered invalidated.
+      transactional @3 :Bool;
+    }
+    timestampNs @4 :Int64;
+    # The sequence order for this event. This is a strictly monotonically
+    # increasing sequence number that is unique within the tail. The onset
+    # event sequence number will always be 0. The purpose of the sequence
+    # is to make it possible to reconstruct the specific ordering of events
+    # in the stream.
+    sequence @5 :UInt32;
+
+    event :union {
+      # When a tail stream is first created, the first event will always be
+      # an onset event.
+      onset @6 :Onset;
+      # The final event in every successfully completed stream be will an outcome
+      # event.
+      outcome @7 :Outcome;
+      # The dropped event is used to identify events that have been dropped from
+      # the stream. The start field indicates the sequence number of the first
+      # event dropped, and the end field indicates the sequence number of the
+      # last event dropped.
+      dropped @8 :Dropped;
+      # Span events mark the ending and outcome of a span.
+      span @9 :Span;
+      # Info events are used at the start of a stage span to identify the kind
+      # of trigger that started the span. For instance, a fetch event will have
+      # a fetch info event at the start of the span.
+      info :union {
+        fetch @10 :FetchEventInfo;
+        jsRpc @11 :JsRpcEventInfo;
+        scheduled @12 :ScheduledEventInfo;
+        alarm @13 :AlarmEventInfo;
+        queue @14 :QueueEventInfo;
+        email @15 :EmailEventInfo;
+        trace @16 :TraceEventInfo;
+        hibernatableWebSocket @17 :HibernatableWebSocketEventInfo;
+        # A custom info event is used to enable arbitrary, non-typed extension
+        # events to be injected. It is most useful as a way of extending
+        # the event stream with new types of events without modifying the
+        # schema. This is a tradeoff. Using a custom event is more flexible
+        # but there's no schema to verify the data.
+        custom @18 :List(Tag);
+      }
+      # Detail events occur throughout a span and may occur many times.
+      detail :union {
```

Comment:
Assuming a model where a stream only ever represents a single request to a single worker, the `Info` event would be expected to occur once per stream, while `Detail` events can occur zero-or-more times. It would also be expected that `Info` events implicitly create a span and that `Detail` events would only ever occur within such a span.

So,  going back to the previous example... let's say we're processing a simple fetch request received by our worker. Let's assume that worker does a console.log and a subrequest. The stream of events would be:

1. Onset (span 0)
2. Info (FetchEventInfo) (starts a span 1)
3. Detail (LogV2) (within span 1)
4. Detail (Subrequest) (within span 1)
5. Detail (SubrequestOutcome) (within span 1)
6. Span (closes span 1)
7. Outcome (closes span 0 and the stream)

We should never have a case where a `Detail` event occurs outside of the span representing the fetch we are processing.